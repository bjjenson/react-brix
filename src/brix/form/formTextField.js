import { useState } from 'react'
import { formatLabel } from './formatLabel'
/**
 * @param  args { import(".").IFormTextFieldArgs}
 */
export const useFormTextField = (args = {}) => {
  const initialValue = args.value || ''
  const requiredMessage = args.requiredMessage || 'Required'

  const [state, setState] = useState({
    value: initialValue,
    error: false,
    helperText: args.helperText,
    pristine: true,
    touched: false,
  })

  const tryValidate = (value, touched) => {
    let result
    if (args.validate) {
      result = args.validate(value, args.name)
    }

    if (!Boolean(result) && !args.optional && touched && value === '') {
      result = requiredMessage
    }

    return touched ? {
      error: Boolean(result),
      helperText: result,
    } : {
        error: state.error,
        helperText: state.helperText,
      }
  }

  const handleChange = ({ target }) => {
    const coercedValue = args.normalize ? args.normalize(target.value) : target.value
    setState({
      ...state,
      value: coercedValue,
      pristine: coercedValue === initialValue,
      ...tryValidate(coercedValue, state.touched),
    })
  }

  const onBlur = () => {
    setState({
      ...state,
      touched: true,
      ...tryValidate(state.value, true),
    })
  }

  const onFocus = () => {
  }

  const validate = () => {
    const result = tryValidate(state.value, true)
    if (result.error) {
      setState({
        ...state,
        ...result,
      })
      return false
    }
    return true
  }

  return {
    props: {
      error: state.error,
      helperText: state.helperText,
      label: formatLabel(args.label, args.optional),
      value: state.value,
      // handlers
      onBlur,
      onChange: handleChange,
      onFocus,
    },
    validate,
  }
}
