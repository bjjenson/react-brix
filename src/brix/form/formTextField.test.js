import { useState } from 'react'
import { useFormTextField } from './formTextField'

jest.mock('react')

const setState = jest.fn()

let initialArgs, event
beforeEach(() => {
  initialArgs = {
    optional: false,
    value: 'the value',
    error: 'on',
    helperText: 'the help',
    label: 'the label',
  }
  event = {
    target: {
      value: '&',
    },
  }

  useState.mockImplementation(a => [a, setState])
})

test('sets initial values', () => {
  useFormTextField(initialArgs)
  expect(useState.mock.calls[0]).toMatchSnapshot()
})

test('defaults value to empty string', () => {
  useFormTextField()
  expect(useState.mock.calls[0]).toMatchSnapshot()
})

test('returns all props needed', () => {
  const actual = useFormTextField(initialArgs)
  expect(actual).toMatchSnapshot()
})

test('adds optional to label if not required', () => {
  const actual = useFormTextField({ label: 'the label', optional: true })
  expect(actual).toMatchSnapshot()
})

describe('validate', () => {
  let validate
  beforeEach(() => {
    validate = jest.fn()
    initialArgs = {
      name: 'field.name',
      validate,
    }
  })

  test('calls validate on change', () => {
    const actual = useFormTextField(initialArgs)
    const e = {
      target: {
        value: '1',
      },
    }
    actual.props.onChange(e)
    expect(validate).toHaveBeenCalledWith('1', initialArgs.name)
  })

  test('does not updates state with results of validation if not "touched"', () => {
    const actual = useFormTextField(initialArgs)
    const e = {
      target: {
        value: '1',
      },
    }
    validate.mockReturnValue('i am error')
    actual.props.onChange(e)
    expect(setState.mock.calls[0]).toMatchSnapshot()
  })

  test('updates state with results of validation if "touched"', () => {
    useState.mockReturnValue([{ touched: true }, setState])
    const actual = useFormTextField(initialArgs)
    const e = {
      target: {
        value: '1',
      },
    }

    validate.mockReturnValue('i am error')
    actual.props.onChange(e)
    expect(setState.mock.calls[0]).toMatchSnapshot()
  })

  test('validates onBlur', () => {
    const actual = useFormTextField({ ...initialArgs, value: '&' })
    actual.props.onBlur()
    expect(validate.mock.calls[0]).toMatchSnapshot()
  })

  test('sets required if empty on blur', () => {
    const actual = useFormTextField({ ...initialArgs, optional: false })
    actual.props.onBlur()
    expect(setState.mock.calls[0]).toMatchSnapshot()
  })

})

test('cancels pristine if initial value changed', () => {
  const actual = useFormTextField(initialArgs)
  actual.props.onChange(event)
  expect(setState.mock.calls[0]).toMatchSnapshot()
})

test('sets touched onBlur', () => {
  const actual = useFormTextField(initialArgs)
  actual.props.onBlur(event)
  expect(setState.mock.calls[0]).toMatchSnapshot()
})
