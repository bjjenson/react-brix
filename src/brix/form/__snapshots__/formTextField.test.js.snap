// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds optional to label if not required 1`] = `
Object {
  "props": Object {
    "error": false,
    "helperText": undefined,
    "label": "the label (optional)",
    "onBlur": [Function],
    "onChange": [Function],
    "onFocus": [Function],
    "value": "",
  },
  "validate": [Function],
}
`;

exports[`cancels pristine if initial value changed 1`] = `
Array [
  Object {
    "error": false,
    "helperText": "the help",
    "pristine": false,
    "touched": false,
    "value": "&",
  },
]
`;

exports[`defaults value to empty string 1`] = `
Array [
  Object {
    "error": false,
    "helperText": undefined,
    "pristine": true,
    "touched": false,
    "value": "",
  },
]
`;

exports[`returns all props needed 1`] = `
Object {
  "props": Object {
    "error": false,
    "helperText": "the help",
    "label": "the label",
    "onBlur": [Function],
    "onChange": [Function],
    "onFocus": [Function],
    "value": "the value",
  },
  "validate": [Function],
}
`;

exports[`sets initial values 1`] = `
Array [
  Object {
    "error": false,
    "helperText": "the help",
    "pristine": true,
    "touched": false,
    "value": "the value",
  },
]
`;

exports[`sets touched onBlur 1`] = `
Array [
  Object {
    "error": false,
    "helperText": undefined,
    "pristine": true,
    "touched": true,
    "value": "the value",
  },
]
`;

exports[`validate does not updates state with results of validation if not "touched" 1`] = `
Array [
  Object {
    "error": false,
    "helperText": undefined,
    "pristine": false,
    "touched": false,
    "value": "1",
  },
]
`;

exports[`validate sets required if empty on blur 1`] = `
Array [
  Object {
    "error": true,
    "helperText": "Required",
    "pristine": true,
    "touched": true,
    "value": "",
  },
]
`;

exports[`validate updates state with results of validation if "touched" 1`] = `
Array [
  Object {
    "error": true,
    "helperText": "i am error",
    "pristine": false,
    "touched": true,
    "value": "1",
  },
]
`;

exports[`validate validates onBlur 1`] = `
Array [
  "&",
  "field.name",
]
`;
